File: 206 Reverse Linked List.py
# -*- coding: utf-8 -*-
"""
Created on Wed Aug  1 16:30:41 2018

@author: zhaoxu

206 Reverse Linked List
"""

# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution(object):
    def reverseList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        
        #Recursive
        if not head or not head.next:
            
            return head
        
        
        
        ln = self.reverseList( head.next )
        
        ln.next = ListNode( head.val )
        
        return ln

#        if not head:
#            return head
        
        

File: CPP Solutions/206 Reverse Linked List.cpp
#include <iostream>
using namespace std;
/**
 * Definition for singly-linked list.
 *  */
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

ListNode* createLinkedList(int arr[], int n){
    if( n == 0 )
        return NULL;

    ListNode* head = new ListNode(arr[0]);
    ListNode* cur = head;
    for(int i = 1; i < n; i++){
        cur->next = new ListNode(arr[i]);
        cur = cur->next;
    }

    return head;
}


void printLinkedList(ListNode* head){
    ListNode* cur = head;
    while(cur != NULL){
        cout << cur->val <<" -> ";
        cur = cur->next;
    }
    cout<<"NULL"<<endl;
    return;
}


void deleteLinkedList(ListNode* head){
    ListNode* cur = head;
    while( cur != NULL ){
        ListNode* delNode = cur;
        cur = cur->next;
        delete delNode;
    }
}

class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        ListNode* pre = NULL;
        ListNode* cur = head;
        while(cur != NULL){
            ListNode* next = cur->next;
            cur->next = pre;
            pre = cur;
            cur = next;
        }
        return pre;
    }
};
int main() {
    int arr[] = {1,2,3,4,5};
    int n = sizeof(arr) / sizeof(int);

    ListNode* head = createLinkedList(arr, n);
    printLinkedList(head);

    ListNode* head2 = Solution().reverseList( head );
    printLinkedList(head2);

    deleteLinkedList(head2);

    return 0;
}